### PR TITLE
Changed neopixel LED API

### PR DESCRIPTION
**Pull request by tridge**
This allows for up to 64 LEDs on a pin, and gives a more flexible API for LED scripting.
Also adds Luminousbee port, and adds layout control over simulated LEDs

### CODE REVIEW

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 1]
```diff
@@ -151,8 +151,9 @@ singleton AP_Vehicle method get_time_flying_ms uint32_t
 
 include AP_SerialLED/AP_SerialLED.h
 singleton AP_SerialLED alias serialLED
-singleton AP_SerialLED method set_num_LEDs boolean uint8_t 1 16 uint8_t 0 32
-singleton AP_SerialLED method set_RGB void uint8_t 1 16 uint32_t 0U UINT32_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX
+singleton AP_SerialLED method set_num_LEDs boolean uint8_t 1 16 uint8_t 0 64
```
Comment (WickedShell) [inline libraries/AP_Scripting/generator/description/bindings.desc]: Ideally the limit of 64 is either a define, or a public static const member of the AP_SerialLED interface and we could then refer to it by name here, which would be much more flexibile if we ever updated the number of LED's supported.

File: libraries/AP_SerialLED/AP_SerialLED.cpp [hunk 2]
```diff
@@ -30,20 +30,28 @@ AP_SerialLED::AP_SerialLED()
...
-void AP_SerialLED::set_RGB(uint8_t chan, uint32_t ledmask, uint8_t red, uint8_t green, uint8_t blue)
+void AP_SerialLED::set_RGB_mask(uint8_t chan, uint32_t ledmask, uint8_t red, uint8_t green, uint8_t blue)
 {
     if (chan >= 1 && chan <= 16) {
         hal.rcout->set_neopixel_rgb_data(chan-1, ledmask, red, green, blue);
     }
 }
 
+// set RGB value on LED number. LED number -1 is all LEDs. First LED is 0. chan is PWM output, 1..16
+void AP_SerialLED::set_RGB(uint8_t chan, int8_t led, uint8_t red, uint8_t green, uint8_t blue)
```
Comment (WickedShell) [inline libraries/AP_SerialLED/AP_SerialLED.cpp]: How is this different from the mask version above? As far as I can tell the implementation actually looks identical, and if this is depending on a argument type difference it's way to close. Was the above mask one not calling `set_neopixel_rgb_data_mask` by accident?

Comment (tridge) [reply libraries/AP_SerialLED/AP_SerialLED.cpp]: I've now removed the mask API. I think it is better to just have the one API, and the -1 for led meaning all LEDs covers most of the use case for mask

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 1 above)
Comment (tridge) [reply libraries/AP_Scripting/generator/description/bindings.desc]: fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 3]
```diff
@@ -1696,37 +1696,46 @@ bool RCOutput::set_neopixel_num_LEDs(const uint16_t chan, uint8_t num_leds)
...
-  and mask of LEDs on the channel
+  and a LED number. LED -1 is all LEDs
 */
-void RCOutput::set_neopixel_rgb_data(const uint16_t chan, uint32_t ledmask, uint8_t red, uint8_t green, uint8_t blue)
+void RCOutput::_set_neopixel_rgb_data(pwm_group *grp, uint8_t idx, uint8_t led, uint8_t red, uint8_t green, uint8_t blue)
+{
+    const uint8_t pad_start_bits = 1;
+    const uint8_t neopixel_bit_length = 24;
+    const uint8_t stride = 4;
+    uint32_t *buf = grp->dma_buffer + (led * neopixel_bit_length + pad_start_bits) * stride + idx;
```
Comment (WickedShell) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Don't you need to sanitize the led number before this somehow? IE what prevents something from calling this with a value of 127?

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 4]
```diff
@@ -151,8 +151,8 @@ singleton AP_Vehicle method get_time_flying_ms uint32_t
 
 include AP_SerialLED/AP_SerialLED.h
 singleton AP_SerialLED alias serialLED
-singleton AP_SerialLED method set_num_LEDs boolean uint8_t 1 16 uint8_t 0 32
-singleton AP_SerialLED method set_RGB void uint8_t 1 16 uint32_t 0U UINT32_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX
+singleton AP_SerialLED method set_num_LEDs boolean uint8_t 1 16 uint8_t 0 AP_SERIALLED_MAX_LEDS
+singleton AP_SerialLED method set_RGB void uint8_t 1 16 int8_t -1 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX
```
Comment (WickedShell) [inline libraries/AP_Scripting/generator/description/bindings.desc]: This is a malformed range check, an `int8_t` can't support a value of `UINT8_MAX`, should be `INT8_MAX`. The generator will secretly fix this for you, but it's worth having right here either way.

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: it is checked in caller in set_neopixel_rgb_data()

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 4 above)
Comment (tridge) [reply libraries/AP_Scripting/generator/description/bindings.desc]: fixed, thanks

**Pull request merged**